v8 Provider Documentation (#123)
MetaMask/metamask-docs: docs/guide/ethereum-provider.md

The presence of the provider object indicates an Ethereum user.

```javascript
// this function detects most providers injected at window.ethereum
import detectEthereumProvider from '@metamask/detect-provider';

const provider = await detectEthereumProvider();

if (provider) {
  // From now on, this should always be true:
  // provider === window.ethereum
  startApp(provider); // initialize your app
} else {
  console.log('Please install MetaMask!');
}
```

::: warning
In Q3 and Q4 of 2020, we are making changes to our provider API that will be breaking for some web3 sites.

Please read the [Upcoming Breaking Changes section](#upcoming-breaking-changes) for details.
:::

The provider API is specified by [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193), and is designed to be minimal.
We recommend that all web3 site developers read the [Basic Usage section](#basic-usage).

## Table of Contents

[[toc]]

## Upcoming Breaking Changes

::: tip
These changes are _upcoming._ Follow [this GitHub issue](https://github.com/MetaMask/metamask-extension/issues/8077) for details.

> If you are new to using the provider, or use [ethers](https://www.npmjs.com/package/ethers), you do not have to worry about these changes, and can skip ahead [to the next section](#api).

Reviewer: >or use ethers

The `chainId` breaking change might still applies for people using `ethers`.

:::

### `window.ethereum` API Changes

In **Q3 2020** (date TBD), we are introducing some breaking changes to this API, which we encourage you to
[read more about here](https://medium.com/metamask/breaking-changes-to-the-metamask-inpage-provider-b4dde069dd0a).

At that time, we will:

- Stop emitting `chainIdChanged`, and instead emit `chainChanged`

> - Ensure that all `chainId` values are **not** 0-prefixed

Reviewer: We should be as specific as we can here. Does this only effect `eth_chainId`? I know that `networkChanged` and `chainChanged`/`chainIdChanged` aren't effected.

I wonder if any data returned by other methods has this incorrectly-prefixed `chainId` in the metadata somewhere 🤔 e.g. `txMeta`

Author: Good point, although I'm not sure where else it may appear. The data returned by e.g. `eth_getTransactionReceipt` and `eth_getTransactionByHash` does not contain the `chainId` (just tested).

Reviewer: I can't see it anywhere else in state either. OK, maybe we just say it's specific to `eth_chainId` for now then, unless we discover otherwise.

```suggestion
- Ensure that chain IDs returned by `eth_chainId` are **not** 0-prefixed
```

  - For example, instead of `0x01`, we will always return `0x1`, wherever the chain ID is returned or accessible.
- Remove the following experimental methods:

>   - `ethereum._metamask.isEnabled`

Reviewer: Why are we removing this? What would recommend as its replacement?

Author: > Why are we removing this?

Several reasons!

- It's sync, but the state is referring to is async
- It refers to a concept called "enabled", which we've done away with
- Its purpose is to keep track of whether accounts are available, but there's `eth_accounts` and `accountsChanged` for that

> What would we recommend as its replacement?

The snippet in "Using the Provider", which is what I advise in 8b8a693

>   - `ethereum._metamask.isApproved`

Reviewer: .... does this exist? 😕 

I don't think we're removing this. I don't think it exists.

Reviewer: Oh, it exists but it's not in prod. Huh. I don't think we need to leave this in for a whole deprecation period if it hasn't made it to prod yet.

Author: wat.

Author: You mean `ethereum._metamask` and its properties, `isEnabled` and `isApproved`?

They do exist in prod, they're just defined in the extension.

Author: I was questioning my sanity there for a second 😅 

![image](https://user-images.githubusercontent.com/25517051/85906644-c5027600-b7c3-11ea-8917-545f6736cec2.png)

Reviewer: 🤦 I was getting this confused for the top-level methods, like `isConnected`. This makes more sense.

Reviewer: I was referencing the older version of the inpage provider, and `createStandardProvider`. I forgot there was additional stuff added in `inpage.js` 🤦

Author: https://youtu.be/NzWizk3e4MM

These changes _may_ break your website.

> Please read our [migration guide](about:blank) for more details.

Reviewer: >`[migration guide](about:blank)`

😅 

I mean... this seems a little bit rude.

Author: lmao

Whoops, that's what I get for not labeling my TODOs. I made a placeholder for this for now and we can handle it in a separate PR.

c6a78b2

### `window.web3` Removal

::: tip
If you do not use the `window.web3` object injected by MetaMask, you will not be affected by these changes.
:::

In **Q4 2020** (date TBD), we will:

- Stop injecting `window.web3` into web pages

> - Remove the `ethereum.autoRefreshOnNetworkChange` property

Reviewer: Are we definitely removing this? I don't think we should mention this until we're prepared to commit to removing this, and we have a plan on how to replace it (e.g. no more automatic refreshes, or a replacement property, etc.)

This leaves me confused about what I'm supposed to do about this removal. Our current error messages strongly advise that all dapps set this property after all.

Author: > Our current error messages strongly advise that all dapps set this property after all.

For the time being, yes, but given that we're removing the object whose behavior the property controls, I don't see how we can keep the property around. I think it'll be confusing if it starts to control the behavior of `window.ethereum` once `window.web3` is removed.

What to do instead can be mentioned in the migration guide. We can have another go at discussing auto-reload today.

Reviewer: It doesn't seem quite accurate to frame the automatic reload as being something the `window.web3` property _does_. I don't see this property as "controlling the behavior of `window.web3`".

I agree that it is confusing to say that we're going to remove `window.web3` without specifying how the automatic reload behavior will change. But still - better to leave the information out if we're not sure yet. Saying we'll remove the property and then deciding to keep it would be _even more_ confusing after all.

I think it is a legitimate possibility that we'll keep the property if we're keeping the automatic reload behavior itself (even if the behavior does change somewhat). We can rename it, but we'd keep the old name around for backwards-compatibility, because... why not. The only scenario where we'd want to completely remove it is if we don't do automatic page reloading anymore at all. Which... seems likely? But we should finalize that decision first in any case. Hopefully we can do that soon.

Author: I'll remove any references to removal for now and we can resume this discussion for the migration guide.

If you rely on the `window.web3` object currently injected by MetaMask, these changes _will_ break your website.
Please read our [migration guide](about:blank) for more details.

## Basic Usage

For any non-trivial Ethereum web application — a.k.a. web3 site — to work, you will have to:

- Detect the Ethereum provider (`window.ethereum`)
- Detect which Ethereum network the user is connected to
- Get the user's Ethereum account(s)

The snippet at the top of this page is sufficient for detecting the provider.
You can learn how to accomplish the other two by reviewing the snippet in the [Using the Provider section](#using-the-provider).

Although any Ethereum operation can be performed via the provider API, most developers use a convenience library with higher-level abstractions.
The most common such libraries are [ethers](https://www.npmjs.com/package/ethers) and [web3.js](https://www.npmjs.com/package/web3).

Even if you use a convenience library, you will likely still need to detect the provider before initializing the library.
Other than that, you can generally learn everything you need to know from the documentation of those libraries, without reading this lower-level API.

> However, for developers of convenience libraries, and for developers who would like to use features that are not yet supported by their favorite libraries, knowledge of the provider API is essential. Read on for more details.

Reviewer: I'm not a fan of how this is framed 🤔 This seems like we're discouraging most devs from using the provider directly, and I don't think we should do that. We should reference convenience libraries, but we don't need to describe them as "high-level" and us as "low-level", and we don't need to make broad generalizations about who might want to use us directly.

Any dev who trusts the EIP-1193 standard to remain stable more than they trust `web3` or `ethers` to remain stable might want to use the provider directly. Or anyone that wants to just use fewer dependencies.

Reviewer: I tried and failed to find a way to re-word this to avoid those problems. Maybe this whole section could be omitted from this page. We already mention convenience libraries in the 'Getting Started' guide.

Author: How about this?
------------------------------------
> The provider API is all you need to create a full-featured web3 application.
> 
> That said, many developers use a convenience library, such as [ethers](https://www.npmjs.com/package/ethers) and [web3.js](https://www.npmjs.com/package/web3), instead of using the provider directly.
> If you are in need of higher-level abstractions than those provided by this API, we recommend that you seek out a convenience library.

Reviewer: That seems fine! We can always improve it later if we think of a better way to frame it.

Author: 482ebb5

  },
];

ethereum
  .request({
    method: 'eth_sendTransaction',
    params,
  })
  .then((result) => {
    // The result varies by by RPC method.
    // For example, this method will return a transaction hash hexadecimal string on success.
  })
  .catch((error) => {
    // If the request fails, the Promise will reject with an error.
  });
```

## Events

The MetaMask provider implements the [Node.js `EventEmitter`](https://nodejs.org/api/events.html) API.
This sections details the events emitted via that API.
There are innumerable `EventEmitter` guides elsewhere, but you can listen for events like this:

```javascript
ethereum.on('accountsChanged', (accounts) => {
  // Handle the new accounts, or lack thereof.
  // "accounts" will always be an array, but it can be empty.
});

ethereum.on('chainChanged', (chainId) => {
  // Handle the new chain.

>   // We recommend reloading the page.

Author: e93e65d

  window.location.reload();
});
```

### connect

```typescript
interface ConnectInfo {
  chainId: string;
}

ethereum.on('connect', handler: (connectInfo: ConnectInfo) => void);
```

The MetaMask provider emits this event when it first becomes able to submit RPC requests to a chain.

> In general, you can assume that the MetaMask provider is connected and has emitted this event by the time you are able to reference it.

Reviewer: I would suggest that for the time being, we not mention this weird behavior. We should recommend that users use `isConnected()` in conjunction with this event handler, so they can check whether the extension is _already_ connected upon page load, and can register the handler to deal with updates.

Though... why is `isConnected()` deprecated as well 🤦‍♂️. I just noticed this.

> In general, you can assume that the MetaMask provider is connected and has emitted this event by the time you are able to reference it.

Author: I believe I addressed this in: 3c76e9c

> In general, you can assume that the MetaMask provider is connected and has emitted this event by the time you are able to reference it.

Reviewer: Yes, I think so. I will create an issue in the `inpage-provider` repo about better representing connectivity.

> In general, you can assume that the MetaMask provider is connected and has emitted this event by the time you are able to reference it.

Reviewer: Here's the issue for improving connectivity: https://github.com/MetaMask/inpage-provider/issues/61

### disconnect

```typescript
ethereum.on('disconnect', handler: (error: ProviderRpcError) => void);
```

The MetaMask provider emits this event if it becomes unable to submit RPC requests to any chain.
In general, this will only happen due to network connectivity issues or some unforeseen error.

Once `disconnect` has been emitted, MetaMask will not accept any new requests until until `connect` is emitted, which may require reloading the page.

### accountsChanged

```typescript
ethereum.on('accountsChanged', handler: (accounts: Array<string>) => void);
```

The MetaMask provider emits this event whenever the return value of the `eth_accounts` RPC method changes.

> `eth_accounts` returns an array that is either empty or contains a single account address that the user decided to expose to the requesting domain (in the case of a website, identified by the _hostname_ of its URL).

Reviewer: I'm not sure this is correct exactly; we assign permissions by origin, not by hostname. We could probably say a little more about which account we're returning too (e.g. that it's the most recently used account).
Maybe something like this?
```suggestion
`eth_accounts` returns an array that is either empty or contains a single account address. The account address returned (if any) is the most recently used account that the requesting site has permission to access. Permissions are granted to _origins_, so all sites on the same origin share the same permissions.
```

Author: cf58561

This means that `accountsChanged` will be emitted whenever the user's exposed account address changes.

::: tip
We plan to allow the `eth_accounts` array to be able to contain multiple addresses in the near future.
:::

### chainChanged

::: warning
**NOTE:** See the [Chain IDs section](#chain-ids) for important information about the MetaMask provider's chain IDs.
:::

```typescript
ethereum.on('chainChanged', handler: (chainId: string) => void);
```

The MetaMask provider emits this event when the currently connected chain changes.

> All RPC requests are be submitted to the currently connected chain.

Reviewer: ```suggestion
All RPC requests are submitted to the currently connected chain.
```

Therefore, it's critical to keep track of the current chain ID by listening for this event.

We _strongly_ recommend reloading the page on chain changes, unless absolutely necessary not to.

```javascript
ethereum.on('chainChanged', (_chainId) => window.location.reload());
```

### message

```typescript
interface ProviderMessage {
  type: string;
  data: unknown;
}

ethereum.on('message', handler: (message: ProviderMessage) => void);
```

The MetaMask provider emits this event when it receives some message that the consumer should be notified of.
The kind of message is identified by the `type` string.

> One prominent example of such messages include updates from RPC subscriptions, such as `eth_subscribe`. For such messages, the message `type` will be `eth_subscription`.

Reviewer: The grammar for this seemed a bit off? It seemed like it was suggesting that `eth_subscription` was the `type` of _any_ "RPC subscriptions" rather than `eth_subscribe` specifically, which is obviously what was meant.
```suggestion
One prominent example is `eth_subscribe` updates. Each update for an `eth_subscribe` RPC subscription will be a message with a `type` of `eth_subscription`.
```

Author: ffa34ef

- Detect which Ethereum network the user is connected to
- Get the user's Ethereum account(s)

<<< @/docs/snippets/handleProvider.js

## Experimental API

::: warning
There is no guarantee that the methods and properties defined in this section will remain stable.
Use it at your own risk.
:::

We expose some experimental, MetaMask-specific methods under the `ethereum._metamask` property.

## Experimental Methods

### ethereum.\_metamask.isApproved() (TO BE REMOVED)

::: danger DANGER
This will be removed in **Q3 2020**.
:::

```typescript
ethereum._metamask.isApproved(): Promise<boolean>;
```

This method returns a `Promise` that resolves to a `boolean` indicating if the caller has access to user accounts.

### ethereum.\_metamask.isEnabled() (TO BE REMOVED)

::: danger DANGER
This will be removed in **Q3 2020**.
:::

```typescript
ethereum._metamask.isEnabled(): boolean;
```

This method returns a `boolean` indicating if the caller has access to user accounts.

### ethereum.\_metamask.isUnlocked()

```typescript
ethereum._metamask.isUnlocked(): Promise<boolean>;
```

This method returns a `Promise` that resolves to a `boolean` indicating if MetaMask is unlocked by the user.

> This is useful for knowing if MetaMask is unlocked in order to provide meaningful instructions to the user during onboarding.

Reviewer: During onboarding specifically? This seems useful generally all of the time, not just during onboarding. The dapp can't do much if the extension is locked.

Author: 3d130b6

Note that this does not indicate if the user has exposed any accounts.

## Legacy API

::: warning
You should **never** rely on any of these methods, properties, or events in practice.
:::

Historically, the Ethereum provider API has been a mess.

> Here, we document the remains of this mess in case you see it used in the wild and wonder what's going on.

Reviewer: 😅 Maybe we should try to tone down the bitterness. Our users don't need to know how difficult a journey this has been.

Author: I want the world to know my pain, Mark!

Author: 3d130b6

## Legacy Properties

> ### ethereum.autoRefreshOnNetworkChange (TO BE REMOVED)

Reviewer: Same note here re: removal - it seems premature.

::: danger DANGER
When `window.web3` is removed in **Q4 2020**, this property will also be removed.

If you don't access `window.web3`, the value of this property will not affect the behavior of your application or MetaMask.

As the consumer of this API, it is your responsbility to handle chain changes using the [`chainChanged` event](#chainChanged).
We recommend reloading the page on `chainChange` or using [ethers](https://www.npmjs.com/package/ethers), which will do so for you.
:::

> By default, this property is `true`.

Reviewer: This isn't quite true. By default this isn't set at all, and we print a console warning asking the dapp to set it.

Author: No, we tell dapps to set it to `false` to silence its related warning, but it is set to `true` by us on initialization.

Reviewer: Huh. Nevermind then. I remembered this incorrectly.

### ethereum.sendAsync() (DEPRECATED)

::: warning
Use [`ethereum.request()`](#ethereum-request-args) instead.
:::

```typescript
interface JsonRpcRequest {
  id: string | undefined;
  jsonrpc: '2.0';
  method: string;
  params?: Array<any>;
}

interface JsonRpcResponse {
  id: string | undefined;
  jsonrpc: '2.0';
  method: string;
  result?: unknown;
  error?: Error;
}

type JsonRpcCallback = (error: Error, response: JsonRpcResponse) => unknown;

ethereum.sendAsync(payload: JsonRpcRequest, callback: JsonRpcCallback): void;
```

This is the ancestor of `ethereum.request`. It only works for JSON-RPC methods, and takes a JSON-RPC request payload object and an error-first callback function as its arguments.

See [the Ethereum JSON-RPC API](https://eips.ethereum.org/EIPS/eip-1474) for details.

### ethereum.send() (DEPRECATED)

::: warning
Use [`ethereum.request()`](#ethereum-request-args) instead.
:::

```typescript
ethereum.send(
  methodOrPayload: string | JsonRpcRequest,
  paramsOrCallback: Array<unknown> | JsonRpcCallback,
): Promise<JsonRpcResponse> | unknown;
```

This method is mostly a Promise-based `sendAsync` with `method` and `params` instead of a payload as arguments.

> However, it behaves in unexpected ways and should be avoided at all costs.

Reviewer: We should probably be a little more specific about how it behaves, embarrassing as it may be. Some poor soul debugging a dapp might thank us for it someday.

Author: GitHub add " 😭" to your reactions you cowards.

Author: 8f424f3

::: danger DANGER
This will be removed in the future.

Please see the [Using the Provider section](#using-the-provider) for the recommended way of keeping track of user accounts.
:::

```typescript
ethereum._metamask.isEnabled(): boolean;
```

This method returns a `boolean` indicating if the caller has access to user accounts.

### ethereum.\_metamask.isUnlocked()

```typescript
ethereum._metamask.isUnlocked(): Promise<boolean>;
```

This method returns a `Promise` that resolves to a `boolean` indicating if MetaMask is unlocked by the user.
MetaMask must be unlocked in order to perform any operation involving user accounts.
Note that this method does not indicate if the user has exposed any accounts to the caller.

## Legacy API

::: warning
You should **never** rely on any of these methods, properties, or events in practice.
:::

This section documents our legacy provider API.
MetaMask only supported this API before the provider API was standardized via [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193) in 2020.
Because of this, you may find web3 sites that use this API, or other providers that implement it.

## Legacy Properties

> ### ethereum.autoRefreshOnNetworkChange (DEPRECATED)

Reviewer: I don't think this should be marked as deprecated yet either. It doesn't make sense to deprecate something we're still actively encouraging people to use. We'd be giving dapp developers contradictory instructions.
```suggestion
### ethereum.autoRefreshOnNetworkChange
```

Author: 882eb8a
